blog/src/a-base32-checksum.md

```diff
@@ -0,0 +1,1242 @@
+      }
+      if (this.test.expected[sub1] === this.input[sub2] && this.test.expected[sub2] === this.input[sub1]) {
+        this.type = (sub2 - sub1 - 1) + '-trans';
+        return;
+      }
+      if (this.test.expected[sub1] === this.test.expected[sub2] && this.input[sub1] === this.input[sub2]) {
+        this.type = (sub2 - sub1 - 1) + '-twin';
+        return;
+      }
+      if (this.test.expected.slice(sub2 + 1) === this.input.slice(sub2 + 1)) {
+        this.type = (sub2 - sub1 - 1) + '-2sub';
+        return;
+      }
+      this.type = 'nsub';
+      return;
+    }
+    this.type = '1sub';
+  }
+
+  validate() {
+    let errors = [];
+    if (typeof this.input !== 'string') {
+      errors.push({code: 'invalid_type', message: 'error input is not a string'});
+    } else if (this.input.length > 200) {
+      errors.push({code: 'invalid_type', message: 'error input is too long'});
+    }
+    errors = errors.concat(this.test.validate());
+    return errors;
+  }
+}
+
+function findSub(expected, input) {
+  for (let i = 0; i < expected.length; i++) {
+    if (expected[i] !== input[i]) {
+      return i;
+    }
+  }
+}
+
+function genChar(alphabet) {
+  return alphabets[alphabet][Math.floor(Math.random() * alphabets[alphabet].length)];
+}
+
+function genPayload(size, alphabet) {
+  let payload = '';
+  for (let i = 0; i < size; i++) {
+    payload += genChar(alphabet);
+  }
+  return payload;
+}
+
+class Test {
+  constructor(test) {
+    this.bits = test.bits || 64;
+    this.alphabet = test.alphabet || 'base10';
+    this.expected = test.expected || this.generate();
+  }
+
+  generate() {
+    const size = symbolCountToEncode(this.bits, this.alphabet);
+    this.expected = genPayload(size, this.alphabet);
+    return this.expected;
+  }
+
+  validate() {
+    const errors = [];
+    if (typeof this.alphabet !== 'string') {
+      errors.push({code: 'invalid_type', message: 'test alphabet is not a string'});
+    } else if (!isValidAlphabet(this.alphabet)) {
+      errors.push({code: 'invalid_type', message: 'test alphabet is unknown'});
+    }
+    if (typeof this.bits !== 'number') {
+      errors.push({code: 'invalid_type', message: 'test bits are not a number'});
+    }
+    if (typeof this.expected !== 'string') {
+      errors.push({code: 'invalid_type', message: 'expected test input is not a string'});
+    } else if (this.expected.length > 200) {
+      errors.push({code: 'invalid_type', message: 'expected test input is too long'});
+    }
+    if (errors.length > 0) { return errors; }
+    const expectedSize = stringSizeToEncode(this.bits, this.alphabet);
+    if (expectedSize !== this.expected.length) {
+      errors.push({code: 'invalid_type', message: 'expected test input does not match alphabet and bits'});
+    }
+    return errors;
+  }
+}
+
+
+  const submission = new Submission({
+    user: '',
+    total: Object.fromEntries(alphabetNames.map(n => [n, 0])),
+    errors: [],
+  });
+
+  const submissionBatchSize = 50;
```

Comment:
Why 50? (Just curious, maybe more people would go through the experiment and reach its conclusion if the sample size took them less time to complete)

Also, once you do reach 50, you still have the text <input> visible, while anything you enters in it doesn't get submitted, so it's a bit confusing at that point. It'd be great to remove the input when the experiment is over, and maybe even replace it with a conclusion as to which algorithm performed better on my sample vs. the whole user entries sample?

Also, maybe the test would be more realistic if the information to copy were not immediately above the input, but to the left of it. Having it right above the input makes it too easy to visually compare the two, which won't be the case in realistic IBAN entry use-cases. What do you think?

https://www.loom.com/share/1052820091f14fe09ea8511dee1d308f

Reply:
Good points!

The batch limit unlocks the overall study results as incentive to hit that mark. Submissions are persisted on the database server on every entry. Even after 50, the new entries are submitted, and the table updated. This can be made clearer; I will add an explanation when we hit 50.